atomic_reactor/util.py

```diff
@@ -1672,3 +1672,207 @@ def chain_get(d, path, default=None):
+        self._custom_base_image = False
+        self._base_from_scratch = False
+        self._source_registry = None
+        self._organization = None
+        self._original_base_image = self._original_parents[-1] if self._original_parents else None
+        self._pullable_parents = []
+        self._local_parents = []
+        self._create_pullable()
+        self._source_and_org_set = False
+
+    @property
+    def original_parents(self):
+        """
+        provides unmodified parent_images from Dockerfile
+        :return: list
+        """
+        return self._original_parents
+
+    @property
+    def original_base_image(self):
+        """
+        provides unmodified base image from Dockerfile
+        :return: str or None
+        """
+        return self._original_base_image
+
+    @property
+    def base_image(self):
+        """
+        provides local tag for base_image, or base image with registry and enclosed
+        when we don't know local tag yet
+        :return: str for scratch or ImageName
+        """
+        if self.base_from_scratch:
+            return SCRATCH_FROM
+
+        if self._local_parents and self._local_parents[0]:
+            return deepcopy(self._local_parents[0])
+        else:
+            return self.base_image_key
+
+    @property
+    def base_image_key(self):
+        """
+        provides base image with registry and enclosed
+        (original base image with added registry and enclosed)
+        :return: str for scratch or ImageName
+        """
+        if self.base_from_scratch:
+            return SCRATCH_FROM
+
+        if self._pullable_parents and self._pullable_parents[0]:
+            return deepcopy(self._pullable_parents[0])
+        else:
+            raise KeyError('base_image')
+
+    @property
+    def base_from_scratch(self):
+        """
+        base image is FROM scratch
+        :return: boolean
+        """
+        return self._base_from_scratch
+
+    @property
+    def custom_base_image(self):
+        """
+        base image is custom (FROM koji/image-build)
+        :return: boolean
+        """
+        return self._custom_base_image
+
+    @property
+    def custom_parent_image(self):
+        """
+        any parent image is custom (FROM koji/image-build)
+        :return: boolean
+        """
+        return self._custom_parent_image
+
+    @property
+    def source_registry(self):
+        return self._source_registry
+
+    @property
+    def organization(self):
+        return self._organization
+
+    def __len__(self):
+        return len(self._pullable_parents)
+
+    def __setitem__(self, key, item):
+        if not isinstance(key, ImageName):
+            key = ImageName.parse(key)
+
+        if not base_image_is_custom(key.to_str()):
+            if not key.registry:
+                key.registry = self._source_registry
+
+            if self._should_enclose(key):
+                key.enclose(self._organization)
```

Comment:
This is repeated in three places, should it be a separate method (e.g. `_enclose_image()`)?

Reply:
yeah I know, but from this it is more clear that it is conditional

_enclose_image is already saying that it will enclose it soo

Reply:
also it is just 2 lines, which we would substitute for 1, so I'd prefer it to be more readable

Reply:
ok, it's fine then